### PR TITLE
Pull documentation of ansible.module_utils from its doc strings.

### PR DESCRIPTION
**Pull request by aknrdureegaesr**
##### SUMMARY

I find the current [`module_utils` - documentation](https://docs.ansible.com/ansible/devel/reference_appendices/module_utils.html) not very helpful. 

The source code itself contains better documentation (even though not perfect).

This pull request activates the documentation from the source code for docs.ansible.com
and improves on that documentation somewhat, in particular, so that the tests pass.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
module_utils

### CODE REVIEW

**Comment by aknrdureegaesr** [comment]
Interestingly, the present pull request is actually building the documentation for the `module_utils` that's installed in the load path of the python that's being used to run the build, not (as was intended) the `module_utils` of the ansible that's being build.

(For the record: The new version of this pull request, commit 456df7000024be4fbaba289f20492200056561b6, fixes that.)

**Comment by aknrdureegaesr** [comment]
The test passes my change on centos 6 but fails it on centos 7. I find this suspicious, as the change itself is documentation-only, so it should not be platform - dependent. Probing into the centos 7 failure, I find (slightly edited for readability):

```
https://download.fedoraproject.org/pub/epel/7/x86_64/repodata/721a6ec0aacec22adcb969c3e421a19ab4805ade557b2e6a6cda3c73ab0253c0-updateinfo.xml.bz2: [Errno 14] HTTP Error 302 - Found
Trying other mirror.
One of the configured repositories failed (EPEL yum repo),
and yum doesn't have enough cached data to continue. At this point the only
safe thing yum can do is fail. ...\n\n 
```

So I presume a flaky test environment. To test this assumption, I'll re-trigger the build by simply re-basing, without any further change, and see what happens.

File: docs/docsite/rst/conf.py [hunk 1]
```diff
@@ -28,6 +28,11 @@
 sys.path.insert(0, os.path.join('ansible', 'lib'))
 sys.path.append(os.path.abspath(os.path.join('..', '_extensions')))
 
+# We want sphinx to document the ansible modules contained in this repository,
+# not those may happen to be installed in the version of Python used to run sphinx.
+# As sphinx loads in order to document, the repository version needs to be loaded:
+sys.path.insert(0, os.path.abspath(os.path.join('..', '..', '..', 'lib')))
+
 VERSION = '2.7'
```
Comment (gundalow) [inline docs/docsite/rst/conf.py:37]: Interesting, would expect this to be 2.8 on `devel`, perhaps it isn't used by anything.

File: docs/docsite/rst/reference_appendices/module_utils.rst [hunk 2]
```diff
@@ -5,65 +5,25 @@
...
 Basic
 ------
-.. glossary::
 
-    ansible.module_utils.basic
-      Utilities in module_utils.basic apply to all module types.
+This functionality is available via ``import ansible.module_utils.basic``:
+
+.. automodule:: ansible.module_utils.basic
+   :members:
```
Comment (gundalow) [inline docs/docsite/rst/reference_appendices/module_utils.rst:27]: What does `members` limit this to?

**Review by gundalow — COMMENTED** [review]
Nice PR. We can look at doing similar for other parts of the code base in the future.

File: docs/docsite/rst/conf.py (hunk 1 above)
Comment (acozine) [reply docs/docsite/rst/conf.py:37]: @gundalow I'm not sure what that does - I updated it (from 2.6) as part of the 2.7 release, thinking perhaps the `VERSION` setting in the Sphinx config should match the `latest` setting on the docsite. We could experiment . . .

File: docs/docsite/rst/reference_appendices/module_utils.rst (hunk 2 above)
Comment (aknrdureegaesr) [reply docs/docsite/rst/reference_appendices/module_utils.rst:27]: "Limit" does not describe it very well... Just saying `automodule::` just gives the doc string of the module itself. I considered it useful to also have the documentation of the module's members, so I added `:members:`. The [sphinx doc](http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) explains this.

File: docs/docsite/rst/conf.py (hunk 1 above)
Comment (aknrdureegaesr) [reply docs/docsite/rst/conf.py:37]: I'm in the same boat with you regarding the "not being sure".

So I left this alone. I think it is not needed for the purpose of this pull request.

Comment (aknrdureegaesr) [reply docs/docsite/rst/conf.py:37]: Being curious, I did an experiment, namenly, setting this VERSION to `3.14`, then running `make clean && make webdocs` and `grep`ing all files for `3.14`. Lessons learned:

"3.14" was not a good choice, as many tests contain an IP 11.12.13.14.

Every single .html - file generated contains a piece of javascript like this:

```
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:'./',
        VERSION:'3.14',
        COLLAPSE_INDEX:false,
        FILE_SUFFIX:'.html',
        HAS_SOURCE:  false
      };
    </script>
```
, checked by
```
$ for f in $(find docs/docsite/_build/html -name \*.html); do grep -q "VERSION:'3.14'" $f || echo $f; done
$
```

**Comment by aknrdureegaesr** [comment]
I don't think you are waiting for anything I should do, or are you?

File: docs/docsite/rst/conf.py (hunk 1 above)
Comment (acozine) [reply docs/docsite/rst/conf.py:37]: Thanks for running the experiment, @aknrdureegaesr. I'll open a separate PR to update the version to 2.8 on `devel`.

**Comment by acozine** [comment]
@aknrdureegaesr we're not waiting on you, no - we might discuss this PR at the docs working group meeting today - starts in five minutes on the #ansible-docs channel on Freenode IRC.

**Comment by aknrdureegaesr** [comment]
In not-so-important news, this PR ended up not being discussed in the [meeting](https://meetbot.fedoraproject.org/ansible-docs/2018-11-20/ansible_documentation_working_group.2018-11-20-15.31.log.html).

**Comment by acozine** [comment]
@aknrdureegaesr very true, would you like to add it properly to next week's agenda? If you can join us, to answer any questions and speak up for your PR, that would be even better!

**Comment by gundalow** [comment]
@acozine would it be possible to have this built on the test docs server?

I like having the docs built from the Python source. It's then a lot easier for people to update the Python files with docs. I'll happily do this for module argspec

**Comment by gundalow** [comment]
Built version of docs is at https://www.famsik.de/ansible/reference_appendices/module_utils

**Comment by gundalow** [comment]
I wonder if example uses of (say) argspec would go into the Python file, or directly in a RST file. Ie would we have a "high level examples page" which links to this new auto generated page "reference material"

**Comment by aknrdureegaesr** [comment]
> Built version of docs is at https://www.famsik.de/ansible/reference_appendices/module_utils

For now, yes. Thanks for linking here @gundalow !

Caveat: That's my private server and a temporary documentation snapshot that will disappear.

**Comment by aknrdureegaesr** [comment]
> @acozine would it be possible to have this built on the test docs server?
> 
> I like having the docs built from the Python source.

For the record: In your local machine, in the top-level `ansible` directory, `make clean && make webdocs` generates the documentation in `docs/docsite/_build/html`. (You probably knew that, but just for the record.)

File: docs/docsite/rst/reference_appendices/module_utils.rst [hunk 3]
```diff
@@ -5,65 +5,25 @@
...
-.. _AnsibleModule:
-.. _ansible.module_utils.basic.AnsibleModule:
 
 AnsibleModule
 --------------
-.. glossary::
 
-    ansible.module_utils.basic.AnsibleModule
-      Utilities in module_utils.basic.AnsibleModule apply to all module types
+This functionality is available via ``from ansible.module_utils.basic import AnsibleModule``.
```
Comment (samccann) [inline docs/docsite/rst/reference_appendices/module_utils.rst]:
```suggestion
To use this functionality, include ``from ansible.module_utils.basic import AnsibleModule`` in your module.
```

Comment (aknrdureegaesr) [reply docs/docsite/rst/reference_appendices/module_utils.rst]: For the record: I had this in two places, fixed in both as of 48e8c74.

**Comment by aknrdureegaesr** [comment]
At this point in time, the result of 48e8c74 is (temporarily) available on my [server](https://www.famsik.de/ansible/reference_appendices/module_utils).

**Comment by aknrdureegaesr** [comment]
# Blocked by broken build

My current plan:

* Wait until bug #49541 has been resolved.

* Trigger another rebuild of this, by doing a rebase to wherever devel then is.

* Assuming the build succeeds, inform people at freenode's `#ansible-doc` that this can be picked up again.

**Comment by aknrdureegaesr** [comment]
# Ready for merge!

The test status has turned to "All checks have passed" without my doing anything. Fine. In my opinion, this thing can and should now be merged into devel. This was agreed upon in the [ansible-docs-meeting day before yesterday](https://meetbot-raw.fedoraproject.org/ansible-docs/2018-12-04/ansible_documentation_working_group_aka_dawgs.2018-12-04-15.31.txt).

ready_for_review

**Comment by aknrdureegaesr** [comment]
I did happen to see an opportunity for improved language in one comment, which I acted upon.

**Comment by aknrdureegaesr** [comment]
A build had been red (undeservedly), we "fixed" that be rerunning without change until green (thank you, @gundalow !), so now it should all be ready for @acozine (or someone else) to have a look and merge. But somehow, this still has a label "needs_revision".  I try to remove this label by:

ready_for_review

Later remark: Doesn't work. Apparently, as a mere contributor, I do not have the power to clean up wrong labels (which comes at no surprise).

**Comment by gundalow** [comment]
@aknrdureegaesr It worked, label was removed.

**Pull request merged**